interpreter/cling/lib/Interpreter/TransactionPool.h

```diff
@@ -73,15 +73,14 @@ namespace cling {
       if (T->getParent())
         T->getParent()->removeNestedTransaction(T);
 
-      T->~Transaction();
-
       // don't overflow the pool
       if (reuse && (m_Transactions.size() < kPoolSize)) {
         T->m_State = Transaction::kNumStates;
+        T->~Transaction();
```

Comment:
> We could set it in the dtor, but I guess this is good enough?

This is good enough if and only if the destructor does not update `m_state`.

> We cannot write to members after ending the object's lifetime, that's UB.

But we are not, are we?  Since the memory as not been released, there is nothing else that (should) write to this object.

Reply:
Object lifetime is ended by dtor. That's not visible in memory but part of the contract with the compiler; it can simply optimize things away as accessing an object after its lifetime ends is UB. As such, setting `m_State` isn't actually useful to cling anyway, as the `Transaction` might get deleted and checking for the deleted `Transaction`'s `m_State` is a memory error.

To answer your question: no, `~Transaction()` does not set `m_State`.

Reply:
#7773 solves this differently.